File: app/src/main/java/fr/athanase/deezer/manager/UserManager.java
package fr.athanase.deezer.manager;

import java.util.List;

import fr.athanase.deezer.database.RealmManager;
import fr.athanase.deezer.network.DeezerClient;
import fr.athanase.deezer.object.jsonobject.PlaylistJson;
import fr.athanase.deezer.object.jsonobject.PlaylistListServiceResponse;
import fr.athanase.deezer.object.jsonobject.UserJson;
import fr.athanase.deezer.object.realmobject.RealmPlaylist;
import fr.athanase.deezer.object.realmobject.RealmUser;
import fr.athanase.deezer.service.DeezerService;
import io.realm.RealmList;
import rx.Observable;


/**
 * Created by vincent on 04/04/2017.
 */

public class UserManager {

    private static UserManager instance;

    private DeezerService service;
    private RealmManager realmManager;

    private UserManager() {
        service = DeezerClient.getInstance().getService();
        realmManager = RealmManager.getInstance();
    }

    public Observable<RealmUser> getUser(int input) {
        Observable<RealmUser> user = realmManager.getUserById(input);
        return user.flatMap(realmUser -> {
            if (realmUser == null) {
                Observable<UserJson> userJson = service.getUser(input);
                        return userJson.filter(userJson1 -> userJson1.getName() != null)
                                .flatMap(userJson1 -> realmManager.insertUser(new RealmUser(userJson1)));
            }
            else {
                return user;
            }
        });
    }

    public Observable<RealmUser> getUserById(int userId) {
        return  RealmManager.getInstance().getUserById(userId);
    }

    public Observable <RealmList<RealmPlaylist>> getUserPlaylists(int userId) {
        Observable <PlaylistListServiceResponse> playlistServiceResponseObservable = service.getUserPlaylist(userId);
        return  playlistServiceResponseObservable
                .flatMap(playlistListServiceResponse -> formatUserPlaylist(playlistListServiceResponse.getData()))
                .flatMap(playlists -> realmManager.updateUserPlaylistData(userId, playlists))
                .flatMap(user -> Observable.just(user.getPlaylists()));
    }

    public Observable <RealmList<RealmPlaylist>> getNextPlaylists(int index, int userId) {
        Observable <PlaylistListServiceResponse> playlistListServiceResponseObservable = service.getNextUserPlaylist(userId, index);
        return playlistListServiceResponseObservable.flatMap(playlistListServiceResponse ->
                formatUserPlaylist(playlistListServiceResponse.getData()));
    }

    public Observable<RealmList<RealmPlaylist>> formatUserPlaylist(List<PlaylistJson> playlistList) {
        RealmList<RealmPlaylist> myList = new RealmList<RealmPlaylist>();
        for (int i = 0; i < playlistList.size(); i++) {
            RealmPlaylist realmPlaylist = new RealmPlaylist(playlistList.get(i));
            myList.add(realmPlaylist);
        }
        return Observable.just(myList);
    }


    public static UserManager getInstance() {
        if (instance != null)
            return instance;
        else {
            instance = new UserManager();
            return instance;
        }
    }
}

File: app/src/main/java/fr/athanase/deezer/manager/UserManager.kt
package fr.athanase.deezer.manager

import fr.athanase.deezer.dao.DeezerQueries
import fr.athanase.deezer.dao.DeezerQueries.Companion.getByIdQuery
import fr.athanase.deezer.dao.fetchSingleItem
import fr.athanase.deezer.dao.saveSingleItemTransaction
import fr.athanase.deezer.dao.updateItemTransaction
import fr.athanase.deezer.model.json.PlaylistJson
import fr.athanase.deezer.model.realm.Playlist
import fr.athanase.deezer.model.realm.User
import fr.athanase.deezer.network.client.UserClient
import io.realm.Realm
import io.realm.RealmList
import rx.Observable

class UserManager {
    companion object {
        fun getUser(userId: Int): Observable<User> {
            val user = Realm.getDefaultInstance().configuration.fetchSingleItem {
                getByIdQuery<User>(userId).invoke(this)
            }

            return user.flatMap { realmUser ->
                if (realmUser == null) {
                    UserClient.getUser(userId).filter { userJson1 -> userJson1.name != null }
                            .flatMap<User> { userJson1 -> Realm.getDefaultInstance().configuration.saveSingleItemTransaction(User(userJson1)) }
                } else {
                    Observable.just(realmUser)
                }
            }
        }

        fun getUserById(userId: Int): Observable<User?> {
            return Realm.getDefaultInstance().configuration.fetchSingleItem {
                DeezerQueries.getByIdQuery<User>(userId).invoke(this)
            }
        }

        fun getUserPlaylists(userId: Int): Observable<RealmList<Playlist>> {
            return UserClient.getUserPlaylist(userId)
                    .flatMap { playlistListServiceResponse -> formatUserPlaylist(playlistListServiceResponse.data) }
                    .flatMap { playlists -> Realm.getDefaultInstance().configuration.updateItemTransaction(DeezerQueries.getByIdQuery<User>(userId)) {
                        it?.playlists?.addAll(playlists)
                        return@updateItemTransaction it
                    }}
                    .flatMap { user ->
                        if (user == null) {
                            return@flatMap Observable.empty<RealmList<Playlist>>()
                        }
                        return@flatMap Observable.just(user.playlists)
                    }
        }
        //.flatMap { playlists -> RealmManager.getInstance().updateUserPlaylistData(userId, playlists) }

        fun getNextPlaylists(index: Int, userId: Int): Observable<RealmList<Playlist>> {
            return UserClient.getNextUserPlaylist(userId, index)
                    .flatMap { playlistListServiceResponse -> formatUserPlaylist(playlistListServiceResponse.data) }
        }

        fun formatUserPlaylist(playlistList: List<PlaylistJson>?): Observable<RealmList<Playlist>> {
            val playlists = playlistList?.let { playlistList } ?: RealmList()
            val myList = RealmList<Playlist>()
            for (i in playlists.indices) {
                val realmPlaylist = Playlist(playlists[i])
                myList.add(realmPlaylist)
            }
            return Observable.just(myList)
        }
    }
}